### ISSUE TITLE
ImportError: cannot import name 'project_config' - docker_and_travis_runs_app branch

### ISSUE THREAD

**Issue by MikeTheCanuck** [hackoregon/team-budget#58]
Currently experiencing ImportError: cannot import name 'project_config' in Travis build:
https://travis-ci.org/hackoregon/team-budget/builds/210124740

When I run the same `./budget_proj/bin/start-proj.sh` script in local computer as the script called by `.travis.yml`, I get the same error.

Running `docker exec -it [container_id] /bin/bash`, I can see that the `project_config.sh` script exists in `/code/budget_proj/`, which can only occur if the `aws s3 cp` command completes successfully.

So the question is, exactly which process is running that causes this final output in the Travis log:

```
  File "/code/budget_proj/settings.py", line 15, in <module>
    from . import project_config
ImportError: cannot import name 'project_config'
```

- is it the `docker-compose -f budget_proj/docker-compose.yml up --build` from `start-proj.sh`?
- is it the `python3 manage.py migrate` in `docker-entrypoint.sh`?
- what is the current working directory of the python(3) executable i.e. do we have to prepend `budget_proj.` for python(3) to find the project_config?  Do we have to change `from .` to some other reference?

**Comment by MikeTheCanuck** [comment]
Results so far:

- tried launching `gunicorn budget_proj.wsgi:application -b :8000' by hand from within `/bin/bash` of the Docker container, no errors encountered

I've added `echo` statements to the beginning of every script to trace which script is executing at exactly what stage of the Travis build.

**Comment by MikeTheCanuck** [comment]
Watching the Docker container build locally:

- configure the local `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` to the creds that have access to the S3 bucket from which `project_config.py` gets downloaded
- run `docker rm $(docker ps -a -q)` and `docker rmi $(docker images -q)` to make sure I'm not using a cached container layer or image
- run `./budget_proj/bin/start-proj.sh`

# Result

```
Step 13/13 : WORKDIR /code
 ---> 9c0b007e67ab
Removing intermediate container 24a52e8fc836
Successfully built 9c0b007e67ab
Creating budgetproj_budget-service_1
Attaching to budgetproj_budget-service_1
budget-service_1  | Running docker-entrypoint.sh...
budget-service_1  | ##############################
budget-service_1  | Running getconfig.sh...
budget-service_1  | CONFIG SETTINGS
budget-service_1  | ##############################
budget-service_1  | PROJ_SETTINGS_DIR budget_proj
budget-service_1  | DEPLOY_TARGET integration
budget-service_1  | CONFIG_BUCKET hacko-budget-config
budget-service_1  | ########################################
budget-service_1  | USING integration CONFIG
budget-service_1  | USING THE hacko-budget-config CONFIG BUCKET
budget-service_1  | ########################################
budget-service_1  | Completed 353 Bytes/353 Bytes (4.5 KiB/s) with 1 file(s) remdownload: s3://hacko-budget-config/integration/project_config.py to budget_proj/project_config.py
budget-service_1  | Operations to perform:
budget-service_1  |   Apply all migrations: admin, auth, budget_app, contenttypes, sessions
budget-service_1  | Running migrations:
budget-service_1  |   No migrations to apply.
budget-service_1  | [2017-03-12 17:42:20 +0000] [18] [INFO] Starting gunicorn 19.7.0
budget-service_1  | [2017-03-12 17:42:20 +0000] [18] [INFO] Listening at: http://0.0.0.0:8000 (18)
budget-service_1  | [2017-03-12 17:42:20 +0000] [18] [INFO] Using worker: sync
```

I browse to http://192.168.99.100:8000/#!/kpm-prod/kpm_prod_list and I can GET the data directly.

# Analysis
This means that the Docker container builds correctly, is able to download the `project_config.py` (which is necessary for the Travis build) and successfully launches the app inside the container, without any **ImportError** or other failures.

I'm going to double-check that all the necessary environment variables are configured in Travis, but otherwise there's nothing else that can possibly vary once you're inside the Docker container, running `docker-entrypoint.sh`.  Previous evidence indicates that `docker-entrypoint.sh` runs, but that the `import . from project_config` statement in `settings.py` fails, which doesn't make sense.

One other thing I noticed, that shouldn't make any difference (since it would affect both the local and Travis build of the Docker container) is that the **exec** permission isn't available for the python scripts inside the container (this test result from running `docker exec -it 016a6ffc416a /bin/bash` on the locally-running container):

```
root@016a6ffc416a:/code/budget_proj# ls -la
total 28
drwxr-xr-x 1 1000 staff  340 Mar 12  2017 .
drwxr-xr-x 1 1000 staff  442 Mar 10 21:49 ..
-rw-r--r-- 1 1000 staff    0 Mar 10 18:48 __init__.py
drwxr-xr-x 1 1000 staff  238 Mar 12  2017 __pycache__
-rw-r--r-- 1 1000 staff  353 Mar 10 02:09 project_config.py
-rw-r--r-- 1 1000 staff  337 Mar 12  2017 project_configBACKUP.py
-rw-r--r-- 1 1000 staff  384 Mar 10 18:48 project_config_template.py
-rw-r--r-- 1 1000 staff 4265 Mar 10 22:32 settings.py
-rw-r--r-- 1 1000 staff 1187 Mar 10 18:48 urls.py
-rw-r--r-- 1 1000 staff  400 Mar 10 18:48 wsgi.py
```

**Comment by MikeTheCanuck** [comment]
One final theory is that the two AWS credential environment variables aren't being passed into the Docker container, so that the **Forbidden** error is simply a matter that the `aws s3 cp` command isn't authenticating to AWS S3 because in the Docker container context, the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables are empty (even though they're populated in the Travis build container context).

If this is true, then that would explain why we keep seeing the **ImportError** in Travis but not locally, *and* the solution to that would be to move the `./bin/getconfig.sh` script execution from `docker-entrypoint.sh` (which runs in the Docker container) to either the `Dockerfile` or to a pre-build step of `docker-compose.yml`.

**Comment by MikeTheCanuck** [comment]
Result: when `.travis.yml` specifies `script: ./budget_proj/bin/start-proj.sh`, the `project_config.py` download proceeds successfully and the app launches:
https://travis-ci.org/hackoregon/team-budget/builds/210307728

```
Step 13 : WORKDIR /code
 ---> Running in 92678c7c0ef8
 ---> 98065d4879ca
Removing intermediate container 92678c7c0ef8
Successfully built 98065d4879ca
Creating budgetproj_budget-service_1
Attaching to budgetproj_budget-service_1
budget-service_1  | Running docker-entrypoint.sh...
budget-service_1  | ##############################
budget-service_1  | Running getconfig.sh...
budget-service_1  | CONFIG SETTINGS
budget-service_1  | ##############################
budget-service_1  | PROJ_SETTINGS_DIR budget_proj
budget-service_1  | DEPLOY_TARGET integration
budget-service_1  | CONFIG_BUCKET hacko-budget-config
budget-service_1  | ########################################
budget-service_1  | USING integration CONFIG
budget-service_1  | USING THE hacko-budget-config CONFIG BUCKET
budget-service_1  | ########################################
download: s3://hacko-budget-config/integration/project_config.py to budget_proj/project_config.py
budget-service_1  | Operations to perform:
budget-service_1  |   Apply all migrations: admin, auth, budget_app, contenttypes, sessions
budget-service_1  | Running migrations:
budget-service_1  |   No migrations to apply.
budget-service_1  | [2017-03-12 18:19:52 +0000] [17] [INFO] Starting gunicorn 19.7.0
budget-service_1  | [2017-03-12 18:19:52 +0000] [17] [INFO] Listening at: http://0.0.0.0:8000 (17)
budget-service_1  | [2017-03-12 18:19:52 +0000] [17] [INFO] Using worker: sync
budget-service_1  | [2017-03-12 18:19:52 +0000] [20] [INFO] Booting worker with pid: 20
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

(The fact that the app doesn't shut itself down, and Travis kills it after a 10min timeout, is irrelevant to this investigation.  This is an expected outcome.)

**Comment by MikeTheCanuck** [comment]
# Result of test-proj.sh

For reasons I can't explain at the moment, inserting `test-proj.sh` into `.travis.yml` fails in a Travis build where `start-proj.sh` succeeds:

```
Step 13 : WORKDIR /code
 ---> Running in 5faaf44a2ee7
 ---> 9a6270afb502
Removing intermediate container 5faaf44a2ee7
Successfully built 9a6270afb502
WARNING: Image for service budget-service was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 316, in execute
    settings.INSTALLED_APPS
  File "/usr/local/lib/python3.5/site-packages/django/conf/__init__.py", line 53, in __getattr__
    self._setup(name)
  File "/usr/local/lib/python3.5/site-packages/django/conf/__init__.py", line 41, in _setup
    self._wrapped = Settings(settings_module)
  File "/usr/local/lib/python3.5/site-packages/django/conf/__init__.py", line 97, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/local/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 673, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/code/budget_proj/settings.py", line 15, in <module>
    from . import project_config
ImportError: cannot import name 'project_config'
The command "./budget_proj/bin/test-proj.sh" exited with 1.
Done. Your build exited with 1.
```

The only notable difference between the two scripts is the parameters on `docker-compose`:

`docker-compose -f budget_proj/docker-compose.yml up --build` 
vs
`docker-compose -f budget_proj/docker-compose.yml run budget-service python manage.py test`

I'm sure there's some subtle reason why **up** behaves as expected and **run [service] python manage.py test** doesn't.  If necessary we can research this in the future, but if the workaround of moving the `getconfig.sh` execution out of Docker context into Travis context works, that'll do for now.

**Comment by MikeTheCanuck** [comment]
I tried moving the call to `getconfig.sh` out of `docker-entrypoint.sh` and into the `before_script` parameter of `.travis.yml`.  (I could have tried adding it somewhere to `docker-compose.yml` but Travis documents the [Build Lifecycle](https://docs.travis-ci.com/user/customizing-the-build/#The-Build-Lifecycle) well, and Docker does not.)

That resulted in the following error:

```
$ ./budget_proj/bin/getconfig.sh
##############################
Running getconfig.sh...
CONFIG SETTINGS
##############################
PROJ_SETTINGS_DIR budget_proj
DEPLOY_TARGET integration
CONFIG_BUCKET hacko-budget-config
########################################
USING integration CONFIG
USING THE hacko-budget-config CONFIG BUCKET
########################################
./budget_proj/bin/getconfig.sh: line 20: aws: command not found
The command "./budget_proj/bin/getconfig.sh" failed and exited with 127 during .
Your build has been stopped.
```

No dice.  OK, we have two options: running the `pip install --upgrade --user awscli` then `getconfig.sh` in Dockerfile, or adding `pip install --upgrade --user awscli` to `getconfig.sh`.

Since we'll need the **awscli** available to Travis to complete the `push.sh` script that deploys the container image from Docker Hub to AWS [either to ECS or Cloud Formation, which uses ECS], let's keep the awscli in the `travis.yml` context.  It's not perfect, but for now the `getconfig.sh` script is a decent place to host that awscli install.

**Comment by MikeTheCanuck** [comment]
**Result**: moving `pip install --upgrade --user awscli` to `getconfig.sh` doesn't address the **ImportError**:

```
ImportError: cannot import name 'project_config'
The command "./budget_proj/bin/test-proj.sh" exited with 1.
Done. Your build exited with 1.
```

This is despite the fact that I see the `download: s3://hacko-budget-config/integration/python_config.py to budget_proj/project_config.py` output at line 301 of this build log (and no error after):
https://travis-ci.org/hackoregon/team-budget/builds/210323621

**Comment by MikeTheCanuck** [comment]
No dice here either - adding `chmod` did nothing for the build outcome:

```
Step 12 : WORKDIR /code
 ---> Running in b7bdbbcb4995
 ---> 462bf5fd8c26
Removing intermediate container b7bdbbcb4995
Successfully built 462bf5fd8c26
WARNING: Image for service budget-service was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 316, in execute
    settings.INSTALLED_APPS
  File "/usr/local/lib/python3.5/site-packages/django/conf/__init__.py", line 53, in __getattr__
    self._setup(name)
  File "/usr/local/lib/python3.5/site-packages/django/conf/__init__.py", line 41, in _setup
    self._wrapped = Settings(settings_module)
  File "/usr/local/lib/python3.5/site-packages/django/conf/__init__.py", line 97, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/local/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 673, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/code/budget_proj/settings.py", line 15, in <module>
    from . import project_config
ImportError: cannot import name 'project_config'
The command "./budget_proj/bin/test-proj.sh" exited with 1.
Done. Your build exited with 1.
```

So again I reiterate: is the problem that the file doesn't exist (download attempted but not completed), can't be accessed (read perms), can't be imported (execute perms??), or that the import isn't looking in the right place (imperfect understanding of CWD and relative path)?

The use of Docker is *supposed* to eliminate such problems (e.g. the application has different dependencies depending on where it runs), and yet I'm plagued by exactly this kind of problem - that apparently the build or execution environment coming from Travis is different than the environment coming from my own laptop.  Apparently Docker is only impervious to these issues once the container is *built* - the lead-up to the image creation is just as messy as ever.

Makes me wonder if we're suffering from some kind of issue like what [this poor soul encountered](https://forums.docker.com/t/permissions-issue-with-compose-build-vs-docker-build/7775) - but sadly no one ever publicly resolved.

**Comment by MikeTheCanuck** [comment]
Reading the `docker-compose run` documentation [here](https://docs.docker.com/compose/reference/run/) I ran across a very important statement:

> First, the command passed by `run` overrides the command defined in the service configuration. For example, if the `web` service configuration is started with `bash`, then `docker-compose run web python app.py` overrides it with `python app.py`.

In our case, with the [current code](https://github.com/hackoregon/team-budget/commit/0d9dfdae9ac508a534f0618ef31d4631573fd27e), that means we're replacing `/code/docker-entrypoint.sh` (from `docker-compose.yml`) with `python manage.py test`.  More explicitly, we're losing this:

```
python3 manage.py migrate
python3 manage.py importcsv budget_app.OCRB /Data/Budget_in_Brief_OCRB_data_All_Years.csv "Source document:source_document" "Service Area:service_area" "Bureau:bureau" "Budget Category:budget_category" "Amount:amount" "FY:fy" "Budget Type:budget_type"
gunicorn budget_proj.wsgi:application -b :8000
```

to get this:

`python manage.py test`

I'm going to try `test-proj.sh` with the `--verbose` parameter on `docker-compose` to see if it can tell me what exactly is happening.

**Comment by MikeTheCanuck** [comment]
OK, this is nuts.

Here's my latest commit, reverting much of the experiments and separating out the build from the test:
https://github.com/hackoregon/team-budget/commit/cbe17c8797fb5fc60ea6900a7523609fdd70309a

And here's the Travis build that, again, failed at the ImportError:
https://travis-ci.org/hackoregon/team-budget/builds/210443297

So to troubleshoot, here's what I've done on my local machine:

- ran `./budget_proj/bin/build-proj.sh`, which ended with `Successfully built 4c8528d213d1`
- ran `docker-compose -f budget_proj/docker-compose.yml start budget-service`, which returned `Starting budget-service ... done`
- ran `docker ps` to get me the CONTAINER ID 917cb3578998
- ran `docker exec -it 917cb3578998 /bin/bash`
- ran `ls -la budget_proj/` which showed this for `project_config.py`: -rw-r--r-- 1 1000 staff  334 Mar 12 18:11 project_config.py  (i.e. `project_config.py` exists, has content and is not executable but is readable by all)
- ran `python manage.py test` which is the essential bit from `test-proj.sh` and it returned:

```
Creating test database for alias 'default'...

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
Destroying test database for alias 'default'...
```

# Analysis

To my understanding, this means that all the files are in place, and the test command will work under simple circumstances.  

It also means that when the test command is launched via `docker-compose -f budget_proj/docker-compose.yml run budget-service python manage.py test`, it will fail even though every prerequisite and every individual command is fine.

**Comment by MikeTheCanuck** [comment]
Dammit, I feel like I'm close now...

I changed `test-proj.sh` to separate out the launch of the container from the execution of the tests:

```
docker-compose -f budget_proj/docker-compose.yml start budget-service
docker-compose -f budget_proj/docker-compose.yml exec budget-service python manage.py test
```

On my local machine that gets this result:

```
./budget_proj/bin/test-proj.sh 
Running test_proj.sh...
Starting budget-service ... done
Creating test database for alias 'default'...

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
Destroying test database for alias 'default'...
```

In Travis, this code gets this result intead:

```
Step 12 : WORKDIR /code
 ---> Running in afedda5ed543
 ---> 9c29a5a0128c
Removing intermediate container afedda5ed543
Successfully built 9c29a5a0128c
Running test_proj.sh...
Starting budget-service ... 
ERROR: No containers to start
ERROR: No container found for budget-service_1
The command "./budget_proj/bin/build-proj.sh && ./budget_proj/bin/test-proj.sh" exited with 1.
```

_aaarrghhhh..._

**Comment by MikeTheCanuck** [comment]
When I combined all the build/start/exec steps into a single script, the maddening result couldn't be more clearly contradictory.  This script:

```
docker-compose -f budget_proj/docker-compose.yml build budget-service
docker-compose -f budget_proj/docker-compose.yml start budget-service
docker-compose -f budget_proj/docker-compose.yml exec budget-service python manage.py test

```

results in:

```
Step 12 : WORKDIR /code
 ---> Running in f2028e3b1914
 ---> bb08dc02f512
Removing intermediate container f2028e3b1914
Successfully built bb08dc02f512
Starting budget-service ... 
ERROR: No containers to start
ERROR: No container found for budget-service_1
```

**Comment by MikeTheCanuck** [comment]
According to [this discussion post](http://www.kuzzfeed.com/thread/fmdwzceo-docker-compose-start-error-no-containers-to-start.html), it might be necessary to use a `create` step to get past the "No containers to start" error.

So for this experiment, my `build-test-proj.sh` script looks like this:

```
docker-compose -f budget_proj/docker-compose.yml build budget-service 
docker-compose -f budget_proj/docker-compose.yml create budget-service 
docker-compose -f budget_proj/docker-compose.yml start budget-service
docker-compose -f budget_proj/docker-compose.yml exec budget-service python manage.py test
```

Result: this arrangement takes us back to the Import Error, so that's no better:

```
ImportError: cannot import name 'project_config'
The command "./budget_proj/bin/build-test-proj.sh" exited with 1.
Done. Your build exited with 1.
```

**Comment by MikeTheCanuck** [comment]
I tried adding some `echo` and `ls -la` statements to see if I could identify what happens when to the container image that is getting reportedly built - here's what I saw:

```
Step 12 : WORKDIR /code
 ---> Running in 150ab8cad566
 ---> 33fd1de7a06d
Removing intermediate container 150ab8cad566
Successfully built 33fd1de7a06d
Build completed...
/home/travis/build/hackoregon/team-budget
total 52
drwxrwxr-x 6 travis travis 4096 Mar 13 17:38 .
drwxrwxr-x 3 travis travis 4096 Mar 13 17:38 ..
drwxrwxr-x 5 travis travis 4096 Mar 13 17:39 budget_proj
drwxrwxr-x 2 travis travis 4096 Mar 13 17:38 Data
-rw-rw-r-- 1 travis travis   53 Mar 13 17:38 .dockerignore
drwxrwxr-x 8 travis travis 4096 Mar 13 17:38 .git
-rw-rw-r-- 1 travis travis  104 Mar 13 17:38 .gitignore
-rw-rw-r-- 1 travis travis 1068 Mar 13 17:38 LICENSE.txt
-rw-rw-r-- 1 travis travis 6055 Mar 13 17:38 README.md
-rw-rw-r-- 1 travis travis  192 Mar 13 17:38 requirements.txt
drwxrwxr-x 2 travis travis 4096 Mar 13 17:38 sql
-rw-rw-r-- 1 travis travis  290 Mar 13 17:38 .travis.yml
Start not executed yet...
Starting budget-service ... 
ERROR: No containers to start
Start completed...
/home/travis/build/hackoregon/team-budget
total 52
drwxrwxr-x 6 travis travis 4096 Mar 13 17:38 .
drwxrwxr-x 3 travis travis 4096 Mar 13 17:38 ..
drwxrwxr-x 5 travis travis 4096 Mar 13 17:39 budget_proj
drwxrwxr-x 2 travis travis 4096 Mar 13 17:38 Data
-rw-rw-r-- 1 travis travis   53 Mar 13 17:38 .dockerignore
drwxrwxr-x 8 travis travis 4096 Mar 13 17:38 .git
-rw-rw-r-- 1 travis travis  104 Mar 13 17:38 .gitignore
-rw-rw-r-- 1 travis travis 1068 Mar 13 17:38 LICENSE.txt
-rw-rw-r-- 1 travis travis 6055 Mar 13 17:38 README.md
-rw-rw-r-- 1 travis travis  192 Mar 13 17:38 requirements.txt
drwxrwxr-x 2 travis travis 4096 Mar 13 17:38 sql
-rw-rw-r-- 1 travis travis  290 Mar 13 17:38 .travis.yml
Exec not started yet...
ERROR: No container found for budget-service_1
Exec completed...
The command "./budget_proj/bin/build-test-proj.sh" exited with 0.
after_success
0.00s$ ./budget_proj/bin/docker-push.sh
/home/travis/build.sh: line 62: ./budget_proj/bin/docker-push.sh: No such file or directory
Done. Your build exited with 0.
```

So between `docker-compose build` and `docker-compose start`, the container that `build` generates can't be found by `start`.  Researching...

**Comment by MikeTheCanuck** [comment]
Additional articles to research, that I just don't want to lose - I'm hopeful they spark ideas beyond their original (seemingly unrelated) symptoms and solutions:

http://stackoverflow.com/questions/34482018/docker-compose-up-does-not-start-a-container
http://stackoverflow.com/questions/31176366/cannot-start-container-on-docker-compose?rq=1
http://stackoverflow.com/questions/41934291/docker-compose-does-not-start-containers?rq=1

**Comment by hassanshamim** [comment]
This is way to much for me to read through, but have you checked that it's downloading project_config to the right directory?

On the backend we discussed moving away from a configuration file containing the secrets, to using environment variables directly to detect application settings (TEST, DEBUG, database type, database location, etc).

If we switched to this I *think* it would be as simple as calling SETENV a few times in the docker config.  Though we still need to keep the secrets somewhere, I think you'd have a lot more flexibility than needing to download a specific file.
